google/resource_dns_record_set.go

```diff
@@ -181,6 +173,29 @@ func resourceDnsRecordSetDelete(d *schema.ResourceData, meta interface{}) error
 
 	zone := d.Get("managed_zone").(string)
 
+	// NS records must always have a value, so we short-circuit delete
+	// this allows terraform delete to work, but may have unexpected
+	// side-effects when deleting just that record set.
+	// Unfortunately, you can set NS records on subdomains, and those
+	// CAN and MUST be deleted, so we need to retrieve the managed zone,
+	// check if what we're looking at is a subdomain, and only not delete
+	// if it's not actually a subdomain
+	var domain string
+	if d.Get("type").(string) == "NS" {
+		if domain == "" {
+			mz, err := config.clientDns.ManagedZones.Get(project, zone).Do()
+			if err != nil {
+				return fmt.Errorf("Error retrieving managed zone %q from %q: %s", zone, project, err)
+			}
+			domain = mz.DnsName
+		}
+
+		if domain == d.Get("name").(string) {
```

Comment:
why not comparing with `record.Name` like above?

Reply:
Above, we're comparing if a record exists with the same name we've configured. Here, we're testing if the root of the domain is the same as what we're trying to delete.

Basically, testing for two different things:

1. On create, we want to make sure any NS records that currently exist are removed in the same request we're adding NS records in.
2. On delete, we want to delete the record iff it's not _both_ an NS record _and_ the same name as the zone.

The first is achieved by listing the NS records on a zone, and making note of any that have the same name as the one we're adding, so they can be removed. The second is achieved by retrieving the name of the managed zone, and comparing it to the record we're trying to delete.